rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +82,48 @@ struct @(spec.base_type.type)_
 {
   using Type = @(spec.base_type.type)_<ContainerAllocator>;
 
-@# constructors (with and without allocator)
-@[for (alloc_type, alloc_name) in [['', ''], ['const ContainerAllocator & ', '_alloc']]]@
-  @('explicit ' if alloc_type else '')@(spec.base_type.type)_(@(alloc_type + alloc_name))
-@# generate initializer lists
-@[  if alloc_name == '_alloc']@
-// *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-@[  end if]@
+  explicit @(spec.base_type.type)_(rosidl_generator_cpp::MessageInitialization init = rosidl_generator_cpp::MessageInitialization::ALL)
+  {
+    if (init == rosidl_generator_cpp::MessageInitialization::ALL) {
+      this->initialize_all();
+    } else if (init == rosidl_generator_cpp::MessageInitialization::SKIP) {
+      // no work to do
+    } else if (init == rosidl_generator_cpp::MessageInitialization::ZERO) {
+      this->zero_initialize();
+    } else if (init == rosidl_generator_cpp::MessageInitialization::DEFAULTS_ONLY) {
+      this->initialize_defaults_only();
+    }
+  }
+
 @{
-used_alloc_name = False
-leading_colon = ':'
+init_list = ":"
+used_alloc = False
 for field in spec.fields:
     # dynamic arrays require allocator if available
-    if alloc_name and field.type.is_array and \
+    if field.type.is_array and \
             (field.type.array_size is None or field.type.is_upper_bound):
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, alloc_name), end='')
-        used_alloc_name = True
-        leading_colon = ' '
-
-    # if default value is available set it in initializer list
-    elif field.type.is_primitive_type() and field.default_value is not None:
-        cpp_value = value_to_cpp(field.type, field.default_value)
-        if leading_colon == ' ':
-            print(',')
-        print('  %s %s(%s)' % (leading_colon, field.name, cpp_value), end='')
-        leading_colon = ' '
-if leading_colon == ' ':
-    print()
+        init_list += ' %s(_alloc),' % (field.name)
```

Comment:
Instead of building a string incrementally I would suggest to collect the initializations here:

```
init_list = []
for ...:
   if ...:
        init_list.append(field.name + '(_alloc)')
```

Reply:
Doesn't `init` need to be passed to the constructor calls of each field?

Reply:
Yeah, good call on both of these points.  Will fix.